File: iosApp/TTCoach/Pose/RtmposeEstimator.swift
// RtmposeEstimator.swift
//
// Stage 2 of the RTMPose pipeline: top-down keypoint estimation via ONNX Runtime.
// Given a person bounding box (image pixels, from `YoloxDetector`) and the camera
// frame, it produces 17 COCO keypoints in ORIGINAL image-pixel coordinates plus a
// per-keypoint score, by running the RTMPose ONNX model and decoding its SimCC output.
//
// All numeric parity math is delegated to `RTMPoseMath` and all pixel sampling to
// `PixelBufferSampler` — this file only wires preprocess -> inference -> postprocess
// per `RTMPOSE_PARITY.md` "Stage 2 — RTMPose".
//
// MATRIX CONVENTION: `PixelBufferSampler.affineWarpToTensor` samples output->source
// (it applies the matrix to each OUTPUT pixel to find the SOURCE read location). rtmlib
// builds the FORWARD warp (`get_warp_matrix(inv=False)`) and lets `cv2.warpAffine`
// invert it internally for sampling. To match that sampling, we pass the INVERSE
// matrix: `RTMPoseMath.getWarpMatrix(..., inverse: true)`. (Confirmed by reading the
// sampler header + its per-output-pixel `affine2x3Apply` loop.)
//
// Execution provider: RTMPose runs fine on the CoreML EP (unlike YOLOX, whose baked-in
// NMS the CoreML EP rejects), so the session is built with the factory default
// `coreML: true` (CPU fallback inside the factory).

import CoreVideo
import Foundation
import OnnxRuntimeBindings
import simd

struct RtmposeEstimator {

    private let session: ORTSession

    /// COCO-17: number of keypoints the model produces.
    static let keypointCount = 17

    /// Use an already-built session.
    init(session: ORTSession) {
        self.session = session
    }

    /// iOS: build a (CoreML-with-CPU-fallback) session from a bundled `.onnx` resource.
    init(bundleResource name: String, in bundle: Bundle = .main) throws {
        self.session = try ORTSessionFactory.makeSession(bundleResource: name, in: bundle, coreML: true)
    }

    /// CLI: build a session from an absolute file path.
    init(modelPath: String) throws {
        self.session = try ORTSessionFactory.makeSession(modelPath: modelPath, coreML: true)
    }

    /// Estimate 17 COCO keypoints for a single person box.
    ///
    /// Returns keypoints in ORIGINAL image-pixel coordinates and per-keypoint scores
    /// (the raw SimCC `val`; NOT clamped here — the backend/CLI clamps to [0,1] at
    /// output per the parity doc). Returns 17 zero keypoints / zero scores on inference
    /// failure (logged) so callers can treat it like a no-detection frame.
    func estimate(
        pixelBuffer: CVPixelBuffer,
        bbox: BoundingBox,
        imageWidth: Int,
        imageHeight: Int
    ) -> (keypoints: [SIMD2<Float>], scores: [Float]) {
        let empty: ([SIMD2<Float>], [Float]) = (
            Array(repeating: SIMD2<Float>(0, 0), count: Self.keypointCount),
            Array(repeating: 0, count: Self.keypointCount)
        )
        do {
            // 1-2. center + aspect-fixed scale (reused for warp AND decode).
            let (center, rawScale) = RTMPoseMath.bboxXyxyToCenterScale(
                x1: bbox.x1, y1: bbox.y1, x2: bbox.x2, y2: bbox.y2, padding: 1.25)
            let scale = RTMPoseMath.topDownAffineScale(
                scale: rawScale, outputW: RTMPoseMath.poseInputW, outputH: RTMPoseMath.poseInputH)

            // 3. inverse (output->source sampling) warp matrix for the sampler.
            let warp = RTMPoseMath.getWarpMatrix(
                center: center, scale: scale, rotDeg: 0,
                outputW: RTMPoseMath.poseInputW, outputH: RTMPoseMath.poseInputH, inverse: true)

            // 4. warp + normalize into the input tensor (BGR, (v-mean)/std).
            let tensor = try PixelBufferSampler.affineWarpToTensor(
                pixelBuffer: pixelBuffer, warpMatrix: warp,
                outputW: RTMPoseMath.poseInputW, outputH: RTMPoseMath.poseInputH,
                meanBGR: RTMPoseMath.meanBGR, stdBGR: RTMPoseMath.stdBGR)

            // 5. run; read simcc_x / simcc_y by name into [[Float]] (K rows each).
            let inputName = try session.inputNames().first ?? "input"
            let outputNames = try session.outputNames()
            let outputs = try session.run(
                withInputs: [inputName: tensor],
                outputNames: Set(outputNames),
                runOptions: nil
            )
            let simccX = try Self.readSimcc(outputs, preferring: "simcc_x", otherAxisName: "simcc_y")
            let simccY = try Self.readSimcc(outputs, preferring: "simcc_y", otherAxisName: "simcc_x")

            // 6-7. argmax decode -> image pixels (aspect-fixed scale + center).
            let (locs, vals) = RTMPoseMath.simccMaximum(simccX: simccX, simccY: simccY)
            let keypoints = RTMPoseMath.decodeKeypoints(
                locs: locs, center: center, scale: scale,
                modelInputW: RTMPoseMath.poseInputW, modelInputH: RTMPoseMath.poseInputH,
                simccSplitRatio: RTMPoseMath.simccSplitRatio)
            return (keypoints, vals)
        } catch {
            print("[RtmposeEstimator] estimate failed: \(error); returning zero keypoints.")
            return empty
        }
    }

    // MARK: - Output reading

    /// Read a SimCC axis output (`simcc_x` or `simcc_y`) into `K` rows of `[Float]`.
    /// Selects the value by exact name first, then by a name containing the axis suffix
    /// (`_x`/`_y`) while excluding the other axis name. Slices rows using the tensor's
    /// own shape `[1, K, W]` (does not hardcode W). Asserts `K == keypointCount`.
    private static func readSimcc(
        _ outputs: [String: ORTValue], preferring name: String, otherAxisName: String
    ) throws -> [[Float]] {
        guard let value = selectAxis(outputs, name: name, otherAxisName: otherAxisName) else {
            throw RtmposeEstimatorError.missingOutput(name)
        }
        let shape = try value.tensorTypeAndShapeInfo().shape.map { $0.intValue }
        guard shape.count == 3 else {
            throw RtmposeEstimatorError.unexpectedShape(name: name, shape: shape)
        }
        let k = shape[1]
        let w = shape[2]
        guard k == keypointCount else {
            throw RtmposeEstimatorError.unexpectedKeypointCount(name: name, k: k)
        }
        let nsData = try value.tensorData()
        let flat = Data(referencing: nsData).withUnsafeBytes { raw in
            Array(raw.bindMemory(to: Float.self))
        }
        guard flat.count >= k * w else {
            throw RtmposeEstimatorError.unexpectedShape(name: name, shape: shape)
        }
        var rows = [[Float]]()
        rows.reserveCapacity(k)
        for i in 0..<k {
            let start = i * w
            rows.append(Array(flat[start..<(start + w)]))
        }
        return rows
    }

    /// Pick the output for one SimCC axis: exact name match, else a name containing the
    /// axis suffix but not the other axis's name (so `simcc_x` doesn't match `simcc_y`).
    private static func selectAxis(
        _ outputs: [String: ORTValue], name: String, otherAxisName: String
    ) -> ORTValue? {
        if let exact = outputs[name] { return exact }
        let suffix = name.hasSuffix("_x") ? "_x" : (name.hasSuffix("_y") ? "_y" : name)
        return outputs.first { key, _ in
            let lower = key.lowercased()
            return lower.contains(suffix) && !lower.contains(otherAxisName.lowercased())
        }?.value
    }
}

enum RtmposeEstimatorError: Error, CustomStringConvertible {
    case missingOutput(String)
    case unexpectedShape(name: String, shape: [Int])
    case unexpectedKeypointCount(name: String, k: Int)

    var description: String {
        switch self {
        case .missingOutput(let n):
            return "RtmposeEstimator: required output '\(n)' not found."
        case .unexpectedShape(let n, let s):
            return "RtmposeEstimator: output '\(n)' has unexpected shape \(s) (want [1, 17, W])."
        case .unexpectedKeypointCount(let n, let k):
            return "RtmposeEstimator: output '\(n)' has K=\(k) keypoints, expected 17."
        }
    }
}

File: app/src/main/java/com/ttcoachai/pose/RtmposeEstimator.kt
package com.ttcoachai.pose

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import ai.onnxruntime.TensorInfo
import android.content.res.AssetManager
import android.graphics.Bitmap
import java.nio.FloatBuffer

/**
 * Stage 2 of the RTMPose pipeline: top-down keypoint estimation via ONNX Runtime. Given a person
 * bounding box (image pixels, from [YoloxDetector]) and the camera frame, it produces 17 COCO
 * keypoints in ORIGINAL image-pixel coordinates plus a per-keypoint score, by running the RTMPose
 * ONNX model and decoding its SimCC output.
 *
 * All numeric parity math is delegated to [RtmposeMath] and all pixel sampling to [BitmapSampler]
 * — this file only wires preprocess -> inference -> postprocess. 1:1 Kotlin port of
 * `iosApp/TTCoach/Pose/RtmposeEstimator.swift`.
 *
 * MATRIX CONVENTION: [BitmapSampler.affineWarpToCHW] samples output->source (it applies the
 * matrix to each OUTPUT pixel to find the SOURCE read location). rtmlib builds the FORWARD warp
 * (`get_warp_matrix(inv=False)`) and lets `cv2.warpAffine` invert it internally for sampling. To
 * match that sampling, we pass the INVERSE matrix: `RtmposeMath.getWarpMatrix(..., inverse =
 * true)`.
 *
 * Execution provider: CPU EP for this slice (unlike iOS, which uses CoreML for pose — Android
 * parity/determinism comes first; NNAPI/ncnn is a later latency swap behind the same seam). Built
 * via [OrtSessionFactory]'s CPU path (`cpuOnly = true`), never NNAPI.
 */
class RtmposeEstimator(private val runner: SimccRunner) : AutoCloseable {

    /** Already-built session/runner (DI/tests). */
    constructor(session: OrtSession) : this(OrtSessionRunner(session))

    /** Android: build a CPU-only session from a bundled `.onnx` asset. */
    constructor(assetManager: AssetManager, assetName: String) : this(
        OrtSessionFactory.makeSession(assetManager, assetName, cpuOnly = true)
    )

    /** Desktop/CLI-style: build a CPU-only session from an absolute file path. */
    constructor(modelPath: String) : this(
        OrtSessionFactory.makeSession(modelPath, cpuOnly = true)
    )

    /**
     * Estimate 17 COCO keypoints for a single person box.
     *
     * Returns keypoints in ORIGINAL image-pixel coordinates and per-keypoint scores (the raw
     * SimCC `val`; NOT clamped here — the backend clamps to [0,1] at output, T5). Returns 17 zero
     * keypoints / zero scores on inference failure (logged) so callers can treat it like a
     * no-detection frame.
     */
    fun estimate(bitmap: Bitmap, bbox: BoundingBox, imageWidth: Int, imageHeight: Int): EstimateResult {
        val pixels = IntArray(bitmap.width * bitmap.height)
        bitmap.getPixels(pixels, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
        return estimate(pixels, bitmap.width, bitmap.height, bbox, imageWidth, imageHeight)
    }

    /**
     * Pure-array core of [estimate] — no `android.graphics.Bitmap`, so it is reachable from plain
     * JUnit tests via a stub [SimccRunner] with no ORT/Robolectric involved.
     */
    fun estimate(
        pixels: IntArray,
        pixelsWidth: Int,
        pixelsHeight: Int,
        bbox: BoundingBox,
        imageWidth: Int,
        imageHeight: Int
    ): EstimateResult {
        val empty = EstimateResult(
            keypoints = List(keypointCount) { Vec2(0f, 0f) },
            scores = FloatArray(keypointCount)
        )
        return try {
            // 1-2. center + aspect-fixed scale (reused for warp AND decode).
            val centerScale = RtmposeMath.bboxXyxyToCenterScale(
                x1 = bbox.x1, y1 = bbox.y1, x2 = bbox.x2, y2 = bbox.y2, padding = 1.25f
            )
            val scale = RtmposeMath.topDownAffineScale(
                scale = centerScale.scale, outputW = RtmposeMath.poseInputW, outputH = RtmposeMath.poseInputH
            )

            // 3. inverse (output->source sampling) warp matrix for the sampler.
            val warp = RtmposeMath.getWarpMatrix(
                center = centerScale.center, scale = scale, rotDeg = 0f,
                outputW = RtmposeMath.poseInputW, outputH = RtmposeMath.poseInputH, inverse = true
            )

            // 4. warp + normalize into the input tensor (BGR, (v-mean)/std).
            val tensor = BitmapSampler.affineWarpToCHW(
                pixels = pixels, width = pixelsWidth, height = pixelsHeight,
                warpMatrix = warp, outputW = RtmposeMath.poseInputW, outputH = RtmposeMath.poseInputH,
                meanBGR = RtmposeMath.meanBGR.toFloatArray(), stdBGR = RtmposeMath.stdBGR.toFloatArray()
            )

            // 5. run; read simcc_x / simcc_y by name into K rows each.
            val outputs = runner.run(tensor)
            val simccX = readSimccAxis(outputs, preferring = "simcc_x", otherAxisName = "simcc_y")
            val simccY = readSimccAxis(outputs, preferring = "simcc_y", otherAxisName = "simcc_x")

            // 6-7. argmax decode -> image pixels (aspect-fixed scale + center).
            val result = RtmposeMath.simccMaximum(simccX, simccY)
            val keypoints = RtmposeMath.decodeKeypoints(
                locs = result.locs, center = centerScale.center, scale = scale,
                modelInputW = RtmposeMath.poseInputW, modelInputH = RtmposeMath.poseInputH,
                simccSplitRatio = RtmposeMath.simccSplitRatio
            )
            EstimateResult(keypoints, result.vals)
        } catch (e: Exception) {
            logWarning("estimate failed: $e; returning zero keypoints.")
            empty
        }
    }

    /** A flat float32 SimCC output axis tensor, `shape = [1, K, W]` row-major. */
    data class SimccAxis(val shape: IntArray, val data: FloatArray)

    /** `keypoints` in ORIGINAL image-pixel coords, `scores` the raw SimCC `val` per keypoint. */
    data class EstimateResult(val keypoints: List<Vec2>, val scores: FloatArray)

    /**
     * The ORT-run seam: given the preprocessed input tensor, returns named SimCC axis outputs.
     * The real `OrtSession.run` cannot run in plain JUnit, so post-processing depends only on
     * this functional interface — tests inject a stub, production injects [OrtSessionRunner].
     */
    fun interface SimccRunner {
        fun run(inputTensor: FloatArray): Map<String, SimccAxis>
    }

    /** Production [SimccRunner]: wraps an [OrtSession], building the `[1,3,256,192]` input tensor. */
    class OrtSessionRunner(private val session: OrtSession) : SimccRunner, AutoCloseable {
        override fun close() {
            session.close()
        }

        override fun run(inputTensor: FloatArray): Map<String, SimccAxis> {
            val env = OrtEnvironment.getEnvironment()
            val inputName = session.inputNames.firstOrNull() ?: "input"
            val shape = longArrayOf(1L, 3L, RtmposeMath.poseInputH.toLong(), RtmposeMath.poseInputW.toLong())
            return OnnxTensor.createTensor(env, FloatBuffer.wrap(inputTensor), shape).use { tensor ->
                session.run(mapOf(inputName to tensor)).use { outputs ->
                    outputs.associate { entry ->
                        val onnxTensor = entry.value as OnnxTensor
                        val shapeInt = (onnxTensor.info as TensorInfo).shape.map { it.toInt() }.toIntArray()
                        val buffer = onnxTensor.floatBuffer
                        val flat = FloatArray(buffer.remaining())
                        buffer.get(flat)
                        entry.key to SimccAxis(shapeInt, flat)
                    }
                }
            }
        }
    }

    /** Closes the underlying runner's owned ONNX session when it is [AutoCloseable] (the
     *  production [OrtSessionRunner]); a no-op for stub runners used in tests. */
    override fun close() {
        (runner as? AutoCloseable)?.close()
    }

    companion object {
        /** COCO-17: number of keypoints the model produces. */
        const val keypointCount = 17

        private const val TAG = "RtmposeEstimator"

        /**
         * Log via `android.util.Log`, swallowing the "not mocked" [RuntimeException] plain JUnit
         * throws (no Android framework stub for `Log` there) so failure-path unit tests can run
         * without Robolectric.
         */
        private fun logWarning(message: String) {
            try {
                android.util.Log.w(TAG, message)
            } catch (_: RuntimeException) {
                // No-op under plain JUnit (Log not mocked); real device/Robolectric logs fine.
            }
        }

        /**
         * Read a SimCC axis output (`simcc_x` or `simcc_y`) into `K` rows of `FloatArray`.
         * Selects the value by exact name first, then by a name containing the axis suffix
         * (`_x`/`_y`) while excluding the other axis name. Slices rows using the tensor's own
         * shape `[1, K, W]` (does not hardcode W). Throws when `K != keypointCount`.
         */
        fun readSimccAxis(
            outputs: Map<String, SimccAxis>, preferring: String, otherAxisName: String
        ): Array<FloatArray> {
            val value = selectAxis(outputs, name = preferring, otherAxisName = otherAxisName)
                ?: throw RtmposeEstimatorError.MissingOutput(preferring)
            val shape = value.shape
            if (shape.size != 3) {
                throw RtmposeEstimatorError.UnexpectedShape(preferring, shape)
            }
            val k = shape[1]
            val w = shape[2]
            if (k != keypointCount) {
                throw RtmposeEstimatorError.UnexpectedKeypointCount(preferring, k)
            }
            val flat = value.data
            if (flat.size < k * w) {
                throw RtmposeEstimatorError.UnexpectedShape(preferring, shape)
            }
            return Array(k) { i ->
                val start = i * w
                flat.copyOfRange(start, start + w)
            }
        }

        /**
         * Pick the output for one SimCC axis: exact name match, else a name containing the axis
         * suffix but not the other axis's name (so `simcc_x` doesn't match `simcc_y`).
         */
        private fun selectAxis(
            outputs: Map<String, SimccAxis>, name: String, otherAxisName: String
        ): SimccAxis? {
            outputs[name]?.let { return it }
            val suffix = when {
                name.endsWith("_x") -> "_x"
                name.endsWith("_y") -> "_y"
                else -> name
            }
            val otherLower = otherAxisName.lowercase()
            return outputs.entries.firstOrNull { (key, _) ->
                val lower = key.lowercase()
                lower.contains(suffix) && !lower.contains(otherLower)
            }?.value
        }
    }
}

/** Errors surfaced when reading SimCC outputs. */
sealed class RtmposeEstimatorError(message: String) : Exception(message) {
    class MissingOutput(name: String) :
        RtmposeEstimatorError("RtmposeEstimator: required output '$name' not found.")

    class UnexpectedShape(name: String, shape: IntArray) :
        RtmposeEstimatorError(
            "RtmposeEstimator: output '$name' has unexpected shape ${shape.toList()} (want [1, 17, W])."
        )

    class UnexpectedKeypointCount(name: String, k: Int) :
        RtmposeEstimatorError("RtmposeEstimator: output '$name' has K=$k keypoints, expected 17.")
}
